File: app/src/main/java/net/opendasharchive/openarchive/ReviewMediaActivity.java
package net.opendasharchive.openarchive;

import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.media.Image;
import android.net.Uri;
import android.os.Bundle;
import android.text.Html;
import android.text.TextUtils;
import android.text.method.LinkMovementMethod;
import android.util.Log;
import android.view.Gravity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import com.github.derlio.waveform.SimpleWaveformView;
import com.github.derlio.waveform.soundfile.SoundFile;
import com.squareup.picasso.Picasso;
import com.stfalcon.frescoimageviewer.ImageViewer;

import net.opendasharchive.openarchive.db.Media;
import net.opendasharchive.openarchive.fragments.MediaViewHolder;
import net.opendasharchive.openarchive.fragments.VideoRequestHandler;
import net.opendasharchive.openarchive.onboarding.FirstStartActivity;
import net.opendasharchive.openarchive.publish.PublishService;
import net.opendasharchive.openarchive.services.PirateBoxSiteController;
import net.opendasharchive.openarchive.services.WebDAVSiteController;
import net.opendasharchive.openarchive.util.FileUtils;
import net.opendasharchive.openarchive.util.Globals;
import net.opendasharchive.openarchive.util.Utility;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SwitchCompat;
import androidx.appcompat.widget.Toolbar;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.content.FileProvider;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import io.scal.secureshareui.controller.ArchiveSiteController;
import io.scal.secureshareui.controller.SiteController;
import io.scal.secureshareui.model.Account;

import static net.opendasharchive.openarchive.MainActivity.INTENT_FILTER_NAME;


public class ReviewMediaActivity extends AppCompatActivity {
    private static String TAG = "ReviewMediaActivity";

    private Context mContext = this;
    private Media mMedia;
    private long currentMediaId;
    private ProgressDialog progressDialog = null;

    private TextView tvTitle;
    private TextView tvDescription;
    private TextView tvAuthor;
    private TextView tvTags;
    private TextView tvLocation;
    private TextView tvUrl;
    private TextView tvLicenseUrl;
    private TextView tvFlagged;


    private SimpleWaveformView swMedia;
    private ImageView ivMedia;

    private SwitchCompat tbDeriv, tbShare, tbComm;

    private MenuItem menuShare;
    private MenuItem menuPublish;


    private ImageView ivEditTags, ivEditLocation, ivEditNotes, ivEditFlag;

    private static Picasso mPicasso;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_review_media);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ActionBar actionbar = getSupportActionBar();
        actionbar.setDisplayHomeAsUpEnabled(true);
     //   actionbar.setHomeAsUpIndicator(R.drawable.arrow_back);

        if (mPicasso == null) {
            VideoRequestHandler videoRequestHandler = new VideoRequestHandler(mContext);

            mPicasso = new Picasso.Builder(mContext)
                    .addRequestHandler(videoRequestHandler)
                    .build();
        }

        // instantiate values
        swMedia = findViewById(R.id.swMedia);

        // display media preview if available
        ivMedia = (ImageView) findViewById(R.id.ivMedia);

        tvTitle = (TextView) findViewById(R.id.tv_title_lbl);
        tvDescription = (TextView) findViewById(R.id.tv_description_lbl);
        tvAuthor = (TextView) findViewById(R.id.tv_author_lbl);
        tvTags = (TextView) findViewById(R.id.tv_tags_lbl);
        tvLocation = (TextView) findViewById(R.id.tv_location_lbl);
        tvUrl = (TextView) findViewById(R.id.tv_url);
        tvFlagged = findViewById(R.id.tv_flag_lbl);


        ivEditTags = findViewById(R.id.ivEditTags);
        ivEditNotes = findViewById(R.id.ivEditNotes);
        ivEditLocation = findViewById(R.id.ivEditLocation);
        ivEditFlag = findViewById(R.id.ivEditFlag);

        tbDeriv = (SwitchCompat)findViewById(R.id.tb_cc_deriv);
        tbDeriv.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                setLicense();
            }
        });
        tbShare = (SwitchCompat)findViewById(R.id.tb_cc_sharealike);
        tbShare.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                setLicense();
            }
        });
        tbComm = (SwitchCompat)findViewById(R.id.tb_cc_comm);
        tbComm.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                setLicense();
            }
        });

        tvLicenseUrl = (TextView) findViewById(R.id.tv_cc_license);

        setTitle("");

        findViewById(R.id.row_flag).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                mMedia.setFlagged(!mMedia.isFlagged());

                updateFlagState();
            }
        });



    }

    private void updateFlagState ()
    {
        ImageView iv = (ImageView)findViewById(R.id.ivEditFlag);

        if (mMedia.isFlagged())
            iv.setImageResource(R.drawable.ic_flag_selected);
        else
            iv.setImageResource(R.drawable.ic_flag_unselected);

        if (mMedia.isFlagged())
            tvFlagged.setText(R.string.status_flagged);
        else
            tvFlagged.setText(R.string.hint_flag);
    }

    private void bindMedia ()
    {

        // set values
        tvTitle.setText(mMedia.getTitle());

        if (!TextUtils.isEmpty(mMedia.getDescription())) {
            tvDescription.setText(mMedia.getDescription());
            ivEditNotes.setImageResource(R.drawable.ic_edit_selected);
        }


        if (!TextUtils.isEmpty(mMedia.getLocation())) {
            tvLocation.setText(mMedia.getLocation());
            ivEditLocation.setImageResource(R.drawable.ic_location_selected);
        }

        if (!TextUtils.isEmpty(mMedia.getTags())) {
            tvTags.setText(mMedia.getTags());
            ivEditTags.setImageResource(R.drawable.ic_tag_selected);
        }

        tvAuthor.setText(mMedia.getAuthor());
        tvLicenseUrl.setText(mMedia.getLicenseUrl());

        if (mMedia.status != Media.STATUS_LOCAL
                && mMedia.status != Media.STATUS_NEW)
        {

            if (mMedia.status == Media.STATUS_UPLOADED||mMedia.status == Media.STATUS_PUBLISHED) {
                tvUrl.setText(Html.fromHtml(getString(R.string.your_media_is_available) + " <a href=\"" + mMedia.getServerUrl() + "\">" + mMedia.getServerUrl() + "</a>"));
                tvUrl.setMovementMethod(LinkMovementMethod.getInstance());
                tvUrl.setVisibility(View.VISIBLE);
            }
            else if (mMedia.status == Media.STATUS_QUEUED) {
                tvUrl.setText("Waiting for upload...");
                tvUrl.setVisibility(View.VISIBLE);
            }
            else if (mMedia.status == Media.STATUS_UPLOADING) {
                tvUrl.setText("Uploading now...");
                tvUrl.setVisibility(View.VISIBLE);
            }

            tvLicenseUrl.setMovementMethod(LinkMovementMethod.getInstance());

            tvTitle.setEnabled(false);

            tvDescription.setEnabled(false);
            if (TextUtils.isEmpty(mMedia.getDescription()))
                tvDescription.setVisibility(View.GONE);

            tvAuthor.setEnabled(false);
            if (TextUtils.isEmpty(mMedia.getAuthor()))
                tvAuthor.setVisibility(View.GONE);

            tvLocation.setEnabled(false);
            if (TextUtils.isEmpty(mMedia.getLocation()))
                tvLocation.setVisibility(View.GONE);

            tvTags.setEnabled(false);
            if (TextUtils.isEmpty(mMedia.getTags()))
                tvTags.setVisibility(View.GONE);

            tvLicenseUrl.setEnabled(false);
            if (TextUtils.isEmpty(mMedia.getLicenseUrl()))
                tvTags.setVisibility(View.GONE);

            findViewById(R.id.groupLicenseChooser).setVisibility(View.GONE);
        }

        if (menuPublish != null) {
            if (mMedia.status == Media.STATUS_LOCAL) {
                menuPublish.setVisible(true);
                menuShare.setVisible(false);

            } else {
                menuShare.setVisible(true);
                menuPublish.setVisible(false);
            }
        }

        updateFlagState ();

    }

    private void setLicense ()
    {

        //the default
        String licenseUrl = "https://creativecommons.org/licenses/by/4.0/";

        if (tbDeriv.isChecked() && tbComm.isChecked() && tbShare.isChecked())
        {
            licenseUrl = "http://creativecommons.org/licenses/by-sa/4.0/";
        }
        else if (tbDeriv.isChecked() && tbShare.isChecked())
        {
            licenseUrl = "http://creativecommons.org/licenses/by-nc-sa/4.0/";
        }
        else if (tbDeriv.isChecked() && tbComm.isChecked())
        {
            licenseUrl = "http://creativecommons.org/licenses/by/4.0/";
        }
        else if (tbDeriv.isChecked())
        {
            licenseUrl = "http://creativecommons.org/licenses/by-nc/4.0/";
        }
        else if (tbComm.isChecked())
        {
            licenseUrl = "http://creativecommons.org/licenses/by-nd/4.0/";
        }

        tvLicenseUrl.setText(licenseUrl);

        mMedia.setLicenseUrl(licenseUrl);
    }

    private void saveMedia ()
    {
        //if deleted
        if (mMedia == null)
            return;

        if (tvTitle.getText().length() > 0)
            mMedia.setTitle(tvTitle.getText().toString());
        else
        {
            //use the file name if the user doesn't set a title
            mMedia.setTitle(new File(mMedia.getOriginalFilePath()).getName());
        }

        mMedia.setDescription(tvDescription.getText().toString());
        mMedia.setAuthor(tvAuthor.getText().toString());
        mMedia.setLocation(tvLocation.getText().toString());
        mMedia.setTags(tvTags.getText().toString());

        setLicense();

        if (mMedia.status == Media.STATUS_NEW)
            mMedia.status = Media.STATUS_LOCAL;

        mMedia.save();
    }

    @Override
    protected void onPause() {
        super.onPause();
        saveMedia();
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {

        getMenuInflater().inflate(R.menu.menu_review_media, menu);

        menuShare = menu.findItem(R.id.menu_item_share);
        menuPublish = menu.findItem(R.id.menu_upload);

        if (mMedia.status != Media.STATUS_UPLOADED)
            menuPublish.setVisible(true);
        else {
            menuShare.setVisible(true);
            menuPublish.setVisible(true);
            menuPublish.setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER);
        }

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // TODO Auto-generated method stub

//        if (drawerListener.onOptionsItemSelected(item)) {
//            return true;
//        }

        Intent intent = null;
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
            case R.id.menu_upload:
                uploadMedia();
                break;
          //  case R.id.menu_item_nearby:
            //    startNearby();
              //  break;
            case R.id.menu_item_share_link:
                shareLink();
                break;
            case R.id.menu_item_share_media:
                shareMedia();
                break;

            case R.id.menu_delete:
                deleteMedia();

                break;

            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private void init() {
        Intent intent = getIntent();

        // get intent extras
        currentMediaId = intent.getLongExtra(Globals.EXTRA_CURRENT_MEDIA_ID, -1);

        // get default metadata sharing values
        SharedPreferences sharedPref = this.getSharedPreferences(Globals.PREF_FILE_KEY, Context.MODE_PRIVATE);

        // check for new file or existing media
        if (currentMediaId >= 0) {
            mMedia = Media.findById(Media.class, currentMediaId);
        } else {
            Utility.toastOnUiThread(this, getString(R.string.error_no_media));
            finish();
            return;
        }

        if (mMedia.getMimeType().startsWith("image")) {

            mPicasso.load(Uri.parse(mMedia.getOriginalFilePath())).fit().centerCrop().into(ivMedia);
        }
        else if (mMedia.getMimeType().startsWith("video")) {

            mPicasso.load(VideoRequestHandler.SCHEME_VIDEO + ":" + mMedia.getOriginalFilePath()).fit().centerCrop().into(ivMedia);
        }
        else if (mMedia.getMimeType().startsWith("audio")) {
            ivMedia.setImageDrawable(getResources().getDrawable(R.drawable.audio_waveform));

            SoundFile soundFile = MediaViewHolder.mSoundFileCache.get( mMedia.getOriginalFilePath());
            if (soundFile != null) {
                swMedia.setAudioFile(soundFile);
                swMedia.setVisibility(View.VISIBLE);
                ivMedia.setVisibility(View.GONE);
            }
        }
        else
            ivMedia.setImageDrawable(getResources().getDrawable(R.drawable.no_thumbnail));

        swMedia.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                showMedia();
            }
        });

        ivMedia.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                showMedia();
            }
        });
    }

    private void showMedia ()
    {

        if (mMedia.mimeType.startsWith("image"))
        {
            ArrayList<Uri> list = new ArrayList<>();
            list.add(Uri.parse(mMedia.getOriginalFilePath()));
            new ImageViewer.Builder(this,list)
                    .setStartPosition(0)
                    .show();
        }
        else {
            /**
            Intent viewMediaIntent = new Intent();
            viewMediaIntent.setAction(Intent.ACTION_VIEW);

            Uri uriFile = FileProvider.getUriForFile(this,
                    BuildConfig.APPLICATION_ID + ".provider",
                    FileUtils.getFile(this,Uri.parse(mMedia.getOriginalFilePath())));

            viewMediaIntent.setDataAndType(uriFile, mMedia.getMimeType());
            viewMediaIntent.putExtra(Intent.EXTRA_STREAM,uriFile);
            viewMediaIntent.addFlags(
                    Intent.FLAG_GRANT_READ_URI_PERMISSION);

            startActivity(viewMediaIntent);
             **/
        }
    }

    private void uploadMedia ()
    {
        
        Account accountWebDAV = new Account(this, WebDAVSiteController.SITE_NAME);
        Account accountArchive = new Account(this, ArchiveSiteController.SITE_NAME);

        /**
        if (PirateBoxSiteController.isPirateBox(this))
        {
            //mark queued
            mMedia.status = Media.STATUS_QUEUED;
            saveMedia();
            bindMedia();
            startService(new Intent(this, PublishService.class));
        }
        else {
         **/
        // if user doesn't have an account
        if (accountWebDAV.isAuthenticated() || accountArchive.isAuthenticated()) {

            //mark queued
            mMedia.status = Media.STATUS_QUEUED;
            saveMedia();
            bindMedia();
            startService(new Intent(this, PublishService.class));
            finish();
        }
        else {
            Intent firstStartIntent = new Intent(this, FirstStartActivity.class);
            startActivity(firstStartIntent);
        }
      //  }

    }

    //share the link to the file on the IA
    private void shareLink ()
    {

        StringBuffer sb = new StringBuffer();
        sb.append("\"").append(mMedia.getTitle()).append("\"").append(' ');
        sb.append(getString(R.string.share_text)).append(' ');
        sb.append(mMedia.getServerUrl());

        Intent sharingIntent = new Intent(android.content.Intent.ACTION_SEND);
        sharingIntent.setType("text/plain");
        sharingIntent.putExtra(android.content.Intent.EXTRA_SUBJECT, mMedia.getTitle());
        sharingIntent.putExtra(android.content.Intent.EXTRA_TEXT, sb.toString());
        startActivity(Intent.createChooser(sharingIntent, getResources().getString(R.string.share_using)));
    }

    //share the link to the file on the IA
    private void shareTorrentLink ()
    {

        StringBuffer sb = new StringBuffer();
        sb.append("\"").append(mMedia.getTitle()).append("\"").append(' ');
        sb.append(getString(R.string.share_torrent_text)).append(' ');

        StringBuffer sbTorrentUrl = new StringBuffer();

        String tagId = Uri.parse(mMedia.getServerUrl()).getLastPathSegment();

        sb.append("https://archive.org/download/");
        sb.append(tagId);
        sb.append("/");
        sb.append(tagId);
        sb.append("_archive.torrent");

        Intent sharingIntent = new Intent(android.content.Intent.ACTION_SEND);
        sharingIntent.setType("text/plain");
        sharingIntent.putExtra(android.content.Intent.EXTRA_SUBJECT, mMedia.getTitle());
        sharingIntent.putExtra(android.content.Intent.EXTRA_TEXT, sb.toString());
        startActivity(Intent.createChooser(sharingIntent, getResources().getString(R.string.share_using)));
    }

    //share the link to the file on the IA
    private void shareMedia ()
    {

        StringBuffer sb = new StringBuffer();
        sb.append("\"").append(mMedia.getTitle()).append("\"").append(' ');
        sb.append(getString(R.string.share_media_text)).append(' ');
        sb.append(mMedia.getServerUrl());

        Intent sharingIntent = new Intent(android.content.Intent.ACTION_SEND);
        sharingIntent.setType(mMedia.getMimeType());
        sharingIntent.putExtra(android.content.Intent.EXTRA_SUBJECT, mMedia.getTitle());
        sharingIntent.putExtra(android.content.Intent.EXTRA_TEXT, sb.toString());

        Uri sharedFileUri = null;
        String mediaPath = mMedia.getOriginalFilePath();
        if (mediaPath.startsWith("content:"))
            sharedFileUri = Uri.parse(mediaPath);
        else
            sharedFileUri = FileProvider.getUriForFile(this, BuildConfig.APPLICATION_ID + ".provider"
                    , new File(mMedia.getOriginalFilePath()));

        sharingIntent.putExtra(Intent.EXTRA_STREAM, sharedFileUri);

        startActivity(Intent.createChooser(sharingIntent, getResources().getString(R.string.share_using)));
    }





    @Override
    protected void onResume() {
        super.onResume();

        init();
        bindMedia();

        LocalBroadcastManager.getInstance(this).registerReceiver(mMessageReceiver,
                new IntentFilter(INTENT_FILTER_NAME));
    }



    private void showSuccess ()
    {
        Toast.makeText(this,getString(R.string.upload_success),Toast.LENGTH_SHORT).show();
    }


    public void showError(final String message) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {

                if(!isFinishing()){
                    new AlertDialog.Builder(ReviewMediaActivity.this)
                            .setTitle("Upload Error")
                            .setMessage(message)
                            .setCancelable(false)
                            .setPositiveButton("ok", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    ReviewMediaActivity.this.finish();
                                }
                            }).create().show();
                }
            }
        });
    }


    @Override
    protected void onDestroy() {
        // Unregister since the activity is about to be closed.
        LocalBroadcastManager.getInstance(this).unregisterReceiver(mMessageReceiver);
        super.onDestroy();
    }


    // Our handler for received Intents. This will be called whenever an Intent
// with an action named "custom-event-name" is broadcasted.
    private BroadcastReceiver mMessageReceiver = new BroadcastReceiver() {

        @Override
        public void onReceive(Context context, Intent intent) {
            // Get extra data included in the Intent
            Log.d("receiver", "Updating media");
            mMedia = Media.findById(Media.class, mMedia.getId());
            bindMedia();

        }
    };

    private void deleteMedia ()
    {


        AlertDialog.Builder build = new AlertDialog.Builder(ReviewMediaActivity.this)
            .setTitle(R.string.menu_delete)
            .setMessage(R.string.alert_delete_media)
            .setCancelable(true).setNegativeButton(R.string.dialog_cancel, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                //do nothing
            }
        })

            .setPositiveButton(R.string.dialog_ok, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {

                    deleteMedia (false, false);
                    finish();

                }
            });


        build.create().show();
    }

    private void deleteRemoteAndLocal ()
    {

    }

    private void deleteMedia (boolean deleteLocalFile, boolean deleteRemoteFile)
    {
        if (deleteRemoteFile)
        {
            mMedia.status = Media.STATUS_DELETE_REMOTE;
            mMedia.save();
            //start upload queue, which will also handle the deletes
            ((OpenArchiveApp)getApplication()).uploadQueue();
        }
        else {
            boolean success = Media.findById(Media.class, currentMediaId).delete();
            Log.d("OAMedia", "Item deleted: " + success);
            mMedia = null;
        }
    }
}

File: app/src/main/java/net/opendasharchive/openarchive/features/media/review/ReviewMediaActivity.kt
package net.opendasharchive.openarchive.features.media.review

import android.content.ActivityNotFoundException
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.work.WorkInfo
import com.bumptech.glide.Glide
import com.squareup.picasso.Picasso
import com.stfalcon.frescoimageviewer.ImageViewer
import net.opendasharchive.openarchive.MainActivity.Companion.INTENT_FILTER_NAME
import net.opendasharchive.openarchive.OpenArchiveApp
import net.opendasharchive.openarchive.R
import net.opendasharchive.openarchive.databinding.ActivityReviewMediaBinding
import net.opendasharchive.openarchive.db.Media
import net.opendasharchive.openarchive.db.MediaViewHolder
import net.opendasharchive.openarchive.db.Space
import net.opendasharchive.openarchive.features.core.BaseActivity
import net.opendasharchive.openarchive.features.onboarding.SpaceSetupActivity
import net.opendasharchive.openarchive.fragments.VideoRequestHandler
import net.opendasharchive.openarchive.util.AlertHelper
import net.opendasharchive.openarchive.util.Prefs
import net.opendasharchive.openarchive.util.extensions.hide
import net.opendasharchive.openarchive.util.extensions.show
import timber.log.Timber
import java.io.File


class ReviewMediaActivity : BaseActivity() {

    companion object {
        const val EXTRA_CURRENT_MEDIA_ID = "archive_extra_current_media_id"
    }

    private lateinit var mBinding: ActivityReviewMediaBinding

    private lateinit var mPicasso: Picasso

    private var mMedia: Media? = null
    private var mMenuPublish: MenuItem? = null
    private var mMenuShare: MenuItem? = null

    private lateinit var mViewModel: ReviewMediaViewModel

    private val mMessageReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            Timber.d("Updating media")
            mMedia = Media.get(mMedia?.id)

            bindMedia()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mBinding = ActivityReviewMediaBinding.inflate(layoutInflater)
        setContentView(mBinding.root)

        mViewModel = ReviewMediaViewModel.getInstance(this, application)

        setSupportActionBar(mBinding.toolbar)
        supportActionBar?.title = ""
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        mPicasso = Picasso.Builder(this)
            .addRequestHandler(VideoRequestHandler(this))
            .build()

        observeValues()
    }

    override fun onResume() {
        super.onResume()

        init()
        bindMedia()

        LocalBroadcastManager.getInstance(this).registerReceiver(
            mMessageReceiver,
            IntentFilter(INTENT_FILTER_NAME))
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_review_media, menu)

        mMenuShare = menu.findItem(R.id.menu_item_share)
        mMenuPublish = menu.findItem(R.id.menu_upload)

        if (mMedia?.sStatus != Media.Status.Uploaded) {
            mMenuPublish?.isVisible = true
        }
        else {
            mMenuShare?.isVisible = true
            mMenuPublish?.isVisible = false
            mMenuPublish?.setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER)
        }

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
            R.id.menu_upload -> checkPermission()
            R.id.menu_item_share_link -> shareLink()
            R.id.menu_item_open_link -> openLink()
            R.id.menu_delete -> showDeleteMediaDialog()
            else -> {}
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onPause() {
        super.onPause()

        LocalBroadcastManager.getInstance(this).unregisterReceiver(mMessageReceiver)

        saveMedia()
    }

    private fun observeValues() {
        mViewModel.workState.observe(this) { workInfo ->
            workInfo.forEach {
                when (it.state) {
                    WorkInfo.State.RUNNING -> {
                        Timber.d("Loading")
                        finish()
                    }

                    WorkInfo.State.SUCCEEDED -> {
                        Timber.d("Succeed")
                    }

                    WorkInfo.State.FAILED -> {
                        Timber.d("Failed")
                    }

                    else -> {
                        Timber.d("workInfo is null")
                    }
                }
            }
        }
    }

    private fun updateFlagState() {
        if (mMedia?.flag == true) {
            mBinding.reviewMetadata.ivEditFlag.setImageResource(R.drawable.ic_flag_selected)
            mBinding.reviewMetadata.tvFlagLbl.setText(R.string.status_flagged)
        }
        else {
            mBinding.reviewMetadata.ivEditFlag.setImageResource(R.drawable.ic_flag_unselected)
            mBinding.reviewMetadata.tvFlagLbl.setText(R.string.hint_flag)
        }

        if (mMedia?.sStatus != Media.Status.Local
            && mMedia?.sStatus != Media.Status.New
            && mMedia?.flag == false)
        {
            mBinding.reviewMetadata.ivEditFlag.hide()
            mBinding.reviewMetadata.tvFlagLbl.hide()
        }
    }

    private fun bindMedia() {

        mBinding.apply {
            reviewMetadata.apply {
                tvTitleLbl.setText(mMedia?.title)

                if (!mMedia?.description.isNullOrEmpty()) {
                    tvDescriptionLbl.setText(mMedia?.description)
                    ivEditNotes.setImageResource(R.drawable.ic_edit_selected)
                }

                if (!mMedia?.location.isNullOrEmpty()) {
                    tvLocationLbl.setText(mMedia?.location)
                    ivEditLocation.setImageResource(R.drawable.ic_location_selected)
                }

                if (!mMedia?.tags.isNullOrEmpty()) {
                    tvTagsLbl.setText(mMedia?.tags)
                    ivEditTags.setImageResource(R.drawable.ic_tag_selected)
                }

                tvAuthorLbl.setText(mMedia?.author)
            }

            if (mMedia?.sStatus != Media.Status.Local
                && mMedia?.sStatus != Media.Status.New
            ) {
                when (mMedia?.sStatus) {
                    Media.Status.Queued -> {
                        tvUrl.text = getString(R.string.waiting_for_upload)
                        tvUrl.show()
                    }

                    Media.Status.Uploading -> {
                        tvUrl.text = getString(R.string.uploading_now)
                        tvUrl.show()
                    }

                    else -> {}
                }

                reviewMetadata.tvTitleLbl.isEnabled = false
                reviewMetadata.tvDescriptionLbl.isEnabled = false

                if (mMedia?.description.isNullOrEmpty()) {
                    reviewMetadata.ivEditNotes.hide()
                    reviewMetadata.tvDescriptionLbl.hint = ""
                }

                reviewMetadata.tvAuthorLbl.isEnabled = false
                reviewMetadata.tvLocationLbl.isEnabled = false

                if (mMedia?.location.isNullOrEmpty()) {
                    reviewMetadata.ivEditLocation.hide()
                    reviewMetadata.tvLocationLbl.hint = ""
                }

                reviewMetadata.tvTagsLbl.isEnabled = false
                if (mMedia?.tags.isNullOrEmpty()) {
                    reviewMetadata.ivEditTags.hide()
                    reviewMetadata.tvTagsLbl.hint = ""
                }
            }
            else {
                reviewMetadata.rowFlag.setOnClickListener {
                    showFirstTimeFlag()
                    mMedia?.flag = !(mMedia?.flag ?: false)
                    updateFlagState()
                }
            }

            if (mMenuPublish != null) {
                if (mMedia?.sStatus == Media.Status.Local) {
                    mMenuPublish?.isVisible = true
                    mMenuShare?.isVisible = false
                } else {
                    mMenuShare?.isVisible = true
                    mMenuPublish?.isVisible = false
                }
            }

            updateFlagState()
        }
    }

    private fun showFirstTimeFlag() {
        if (Prefs.flagHintShown) return

        AlertHelper.show(this, R.string.popup_flag_desc, R.string.popup_flag_title)

        Prefs.flagHintShown = true
    }

    private fun saveMedia() {
        //if deleted
        mMedia?.title = if (mBinding.reviewMetadata.tvTitleLbl.text.isNotEmpty()) {
            mBinding.reviewMetadata.tvTitleLbl.text.toString()
        } else {
            // Use the file name if the user doesn't set a title.
            File(mMedia?.originalFilePath ?: "").name
        }

        mMedia?.description = mBinding.reviewMetadata.tvDescriptionLbl.text.toString()
        mMedia?.author = mBinding.reviewMetadata.tvAuthorLbl.text.toString()
        mMedia?.location = mBinding.reviewMetadata.tvLocationLbl.text.toString()
        mMedia?.tags = mBinding.reviewMetadata.tvTagsLbl.text.toString()
        mMedia?.licenseUrl = mMedia?.project?.licenseUrl

        if (mMedia?.sStatus == Media.Status.New) mMedia?.sStatus = Media.Status.Local

        mMedia?.save()
    }

    private fun init() {

        val intent = intent

        // get intent extras
        val mediaId = intent.getLongExtra(EXTRA_CURRENT_MEDIA_ID, -1)

        // check for new file or existing media
        if (mediaId >= 0) {
            mMedia = Media.get(mediaId)
        }
        else {
            runOnUiThread {
                Toast.makeText(applicationContext, R.string.error_no_media, Toast.LENGTH_SHORT).show()
            }

            finish()

            return
        }

        if (mMedia?.mimeType?.startsWith("image") == true) {
            Glide.with(mBinding.ivMedia.context)
                .load(Uri.parse(mMedia?.originalFilePath))
                .fitCenter()
                .into(mBinding.ivMedia)
        }
        else if (mMedia?.mimeType?.startsWith("video") == true) {
            mPicasso.load(VideoRequestHandler.SCHEME_VIDEO + ":" + mMedia?.originalFilePath)
                ?.fit()
                ?.centerCrop()
                ?.into(mBinding.ivMedia)
        }
        else if (mMedia?.mimeType?.startsWith("audio") == true) {
            mBinding.ivMedia.setImageDrawable(
                ContextCompat.getDrawable(this,R.drawable.audio_waveform))

            val soundFile = MediaViewHolder.mSoundFileCache[mMedia?.originalFilePath]
            if (soundFile != null) {
                mBinding.swMedia.setAudioFile(soundFile)
                mBinding.swMedia.show()
                mBinding.ivMedia.hide()
            }
        }
        else {
            mBinding.ivMedia.setImageDrawable(
                ContextCompat.getDrawable(this, R.drawable.no_thumbnail))
        }

        mBinding.swMedia.setOnClickListener { showMedia() }
        mBinding.ivMedia.setOnClickListener { showMedia() }
    }

    private fun showMedia() {
        if (mMedia?.mimeType?.startsWith("image") == true) {
            ImageViewer.Builder(this, listOf(Uri.parse(mMedia?.originalFilePath)))
                .setStartPosition(0)
                .show()
        }
    }

    private fun checkPermission() {
        if (Space.current != null) {
            // mark queued
            mMedia?.sStatus = Media.Status.Queued
            saveMedia()
            bindMedia()
            mViewModel.applyMedia()
        }
        else {
            startActivity(Intent(this, SpaceSetupActivity::class.java))
        }
    }

    // Share the link to the file on the IA.
    private fun shareLink() {
        val intent = Intent(Intent.ACTION_SEND)
        intent.type = "text/plain"
        intent.putExtra(Intent.EXTRA_SUBJECT, mMedia?.title)
        intent.putExtra(Intent.EXTRA_TEXT,
            "\"${mMedia?.title}\" ${getString(R.string.share_text)} ${mMedia?.serverUrl}")

        startActivity(Intent.createChooser(intent, resources.getString(R.string.share_using)))
    }

    private fun openLink() {
        try {
            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(mMedia?.serverUrl)))
        }
        catch (e: ActivityNotFoundException) {
            Toast.makeText(
                this, "No application can handle this request."
                        + " Please install a webbrowser", Toast.LENGTH_LONG
            ).show()
            e.printStackTrace()
        }
    }

    private fun showDeleteMediaDialog() {
        AlertHelper.show(
            this,
            R.string.popup_remove_desc,
            R.string.popup_remove_title,
            buttons = listOf(
                AlertHelper.negativeButton(),
                AlertHelper.positiveButton { _, _ ->
                    deleteMedia()
                }))
    }

    private fun deleteMedia() {
        if (!mMedia?.serverUrl.isNullOrEmpty() || mMedia?.sStatus == Media.Status.Uploaded || mMedia?.sStatus == Media.Status.Published) {
            mMedia?.sStatus = Media.Status.DeleteRemote
            mMedia?.save()

            // Start upload queue, which will also handle the deletes.
            (application as OpenArchiveApp).startUploadService()

            finish()
        }
        else {
            val success = mMedia?.delete()

            Timber.d("Item deleted: $success")

            finish()
        }
    }
}
